oprf/oprf.go

```diff
@@ -0,0 +1,238 @@
+package oprf
+
+import (
+	"encoding/binary"
+	"errors"
+	"fmt"
+
+	"github.com/cloudflare/circl/oprf/group"
+)
+
+const (
+	// OPRFCurve25519 is the constant to represent the OPRF curve25519 with SHA-512 (ELL2-RO) group.
+	OPRFCurve25519 uint16 = 0x0001
+	// OPRFCurve448 is the constant to represent the OPRF curve448 with SHA-512 (ELL2-RO) group.
+	OPRFCurve448 uint16 = 0x0002
+	// OPRFP256 is the constant to represent the OPRF P-256 with SHA-512 (SSWU-RO) group.
+	OPRFP256 uint16 = 0x0003
+	// OPRFP384 is the constant to represent the OPRF P-384 with SHA-512 (SSWU-RO) group.
+	OPRFP384 uint16 = 0x0004
+	// OPRFP521 is the constant to represent the OPRF P-521 with SHA-512 (SSWU-RO) group.
+	OPRFP521 uint16 = 0x0005
+)
+
+const (
+	// OPRFMode is the context string to define a OPRF.
+	OPRFMode string = "000"
+)
+
+var (
+	// ErrUnsupportedGroup is an error stating that the ciphersuite chosen is not supported
+	ErrUnsupportedGroup = errors.New("the chosen group is not supported")
+)
+
+// BlindToken corresponds to a token that has been blinded.
+// Internally, it is a serialized Element.
+type BlindToken []byte
+
+// IssuedToken corresponds to a token that has been issued.
+// Internally, it is a serialized Element.
+type IssuedToken []byte
+
+// Token is the object issuance of the protocol.
+type Token struct {
+	data  []byte
+	blind *group.Scalar
+}
+
+// Evaluation corresponds to the evaluation over a token.
+type Evaluation struct {
+	element []byte
+}
+
+// KeyPair is an struct containing a public and private key.
+type KeyPair struct {
+	PubK  *group.Element
+	PrivK *group.Scalar
+}
+
+// Client is a representation of a Client during protocol execution.
+type Client struct {
+	suite *group.Ciphersuite
+	ctx   string
+}
+
+// ClientContext implements the functionality of a Client.
+type ClientContext interface {
+	// Blind generates a token and blinded data.
+	Blind(in []byte) (*Token, *BlindToken, error)
+
+	// Unblind unblinds the server response.
+	Unblind(t *Token, e *Evaluation) (IssuedToken, error)
+
+	// Finalize outputs a byte array that corresponds to its input.
+	Finalize(t *Token, issuedT IssuedToken, info []byte) []byte
```

Comment:
Do these need to be two separate functions? Or can we merge them, e.g., into an API like so:

```
type ClientRequestContext interface {
   Finalize(e *Evaluation, info []byte) ([]byte, error)
}

type ClientContext interface {
   Request(in []byte) (*BlindToken, ClientRequestContext)
}
```

Clients would then use this like so:

```
input := ... // []byte
token, requestContext := client.Request(input)
// send token, get back `evaluation`
output, err := requestContext.Finalize(evaluation, []byte("info"))
// process err and output
```

I think this would be less error prone, since the implementation takes care of the OPRF details.

Reply:
Ok, so I think I'm going to clarify here what i think this actually means. This is referring to the 3 methods of the ClientContext not only the 2 last ones. The idea is to change:

```
	// Blind generates a token and blinded data.
	Blind(in []byte) (*Token, *BlindToken, error)
```

by 

```
   Request(in []byte) (*Token, *BlindToken, ClientRequestContext)
```

and:

```
	Unblind(t *Token, e *Evaluation) (IssuedToken, error)

	// Finalize outputs a byte array that corresponds to its input.
	Finalize(t *Token, issuedT IssuedToken, info []byte) []byte
```

by 

```
type ClientRequestContext interface {
   Finalize(t *Token, e *Evaluation, info []byte) ([]byte, error)
}
```

is it correct? 

Note that `t *Token` is still needed..

Reply:
I think I addressed this in a simpler way. But let me know what you think.